File: lib/js/bs_MapString.js
'use strict';

var Caml_primitive = require("./caml_primitive.js");
var Bs_internalAVLtree = require("./bs_internalAVLtree.js");

function add(t, x, data) {
  if (t !== null) {
    var l = t.left;
    var k = t.key;
    var v = t.value;
    var r = t.right;
    if (x === k) {
      return {
              left: l,
              key: x,
              value: data,
              right: r,
              h: t.h
            };
    } else if (x < k) {
      return Bs_internalAVLtree.bal(add(l, x, data), k, v, r);
    } else {
      return Bs_internalAVLtree.bal(l, k, v, add(r, x, data));
    }
  } else {
    return {
            left: null,
            key: x,
            value: data,
            right: null,
            h: 1
          };
  }
}

function findOpt(x, _n) {
  while(true) {
    var n = _n;
    if (n !== null) {
      var v = n.key;
      if (x === v) {
        return /* Some */[n.value];
      } else {
        _n = x < v ? n.left : n.right;
        continue ;
        
      }
    } else {
      return /* None */0;
    }
  };
}

function findAssert(x, _n) {
  while(true) {
    var n = _n;
    if (n !== null) {
      var v = n.key;
      if (x === v) {
        return n.value;
      } else {
        _n = x < v ? n.left : n.right;
        continue ;
        
      }
    } else {
      throw new Error("Not_found");
    }
  };
}

function findWithDefault(_n, x, def) {
  while(true) {
    var n = _n;
    if (n !== null) {
      var v = n.key;
      if (x === v) {
        return n.value;
      } else {
        _n = x < v ? n.left : n.right;
        continue ;
        
      }
    } else {
      return def;
    }
  };
}

function mem(_n, x) {
  while(true) {
    var n = _n;
    if (n !== null) {
      var v = n.key;
      if (x === v) {
        return /* true */1;
      } else {
        _n = x < v ? n.left : n.right;
        continue ;
        
      }
    } else {
      return /* false */0;
    }
  };
}

function remove(n, x) {
  if (n !== null) {
    var l = n.left;
    var v = n.key;
    var r = n.right;
    if (x === v) {
      if (l !== null) {
        if (r !== null) {
          var kr = [r.key];
          var vr = [r.value];
          var r$1 = Bs_internalAVLtree.removeMinAuxWithRef(r, kr, vr);
          return Bs_internalAVLtree.bal(l, kr[0], vr[0], r$1);
        } else {
          return l;
        }
      } else {
        return r;
      }
    } else if (x < v) {
      return Bs_internalAVLtree.bal(remove(l, x), v, n.value, r);
    } else {
      return Bs_internalAVLtree.bal(l, v, n.value, remove(r, x));
    }
  } else {
    return n;
  }
}

function splitAux(x, n) {
  var l = n.left;
  var v = n.key;
  var d = n.value;
  var r = n.right;
  if (x === v) {
    return /* tuple */[
            l,
            /* Some */[d],
            r
          ];
  } else if (x < v) {
    if (l !== null) {
      var match = splitAux(x, l);
      return /* tuple */[
              match[0],
              match[1],
              Bs_internalAVLtree.join(match[2], v, d, r)
            ];
    } else {
      return /* tuple */[
              null,
              /* None */0,
              n
            ];
    }
  } else if (r !== null) {
    var match$1 = splitAux(x, r);
    return /* tuple */[
            Bs_internalAVLtree.join(l, v, d, match$1[0]),
            match$1[1],
            match$1[2]
          ];
  } else {
    return /* tuple */[
            n,
            /* None */0,
            null
          ];
  }
}

function split(x, n) {
  if (n !== null) {
    return splitAux(x, n);
  } else {
    return /* tuple */[
            null,
            /* None */0,
            null
          ];
  }
}

function merge(s1, s2, f) {
  var exit = 0;
  if (s1 !== null) {
    if (s1.h >= (
        s2 !== null ? s2.h : 0
      )) {
      var l1 = s1.left;
      var v1 = s1.key;
      var d1 = s1.value;
      var r1 = s1.right;
      var match = split(v1, s2);
      return Bs_internalAVLtree.concatOrJoin(merge(l1, match[0], f), v1, f(v1, /* Some */[d1], match[1]), merge(r1, match[2], f));
    } else {
      exit = 1;
    }
  } else if (s2 !== null) {
    exit = 1;
  } else {
    return null;
  }
  if (exit === 1) {
    if (s2 !== null) {
      var l2 = s2.left;
      var v2 = s2.key;
      var d2 = s2.value;
      var r2 = s2.right;
      var match$1 = split(v2, s1);
      return Bs_internalAVLtree.concatOrJoin(merge(match$1[0], l2, f), v2, f(v2, match$1[1], /* Some */[d2]), merge(match$1[2], r2, f));
    } else {
      return /* assert false */0;
    }
  }
  
}

function cmp(s1, s2, cmp$1) {
  var len1 = Bs_internalAVLtree.length0(s1);
  var len2 = Bs_internalAVLtree.length0(s2);
  if (len1 === len2) {
    var _e1 = Bs_internalAVLtree.stackAllLeft(s1, /* [] */0);
    var _e2 = Bs_internalAVLtree.stackAllLeft(s2, /* [] */0);
    var vcmp = cmp$1;
    while(true) {
      var e2 = _e2;
      var e1 = _e1;
      if (e1) {
        if (e2) {
          var h2 = e2[0];
          var h1 = e1[0];
          var c = Caml_primitive.caml_string_compare(h1.key, h2.key);
          if (c) {
            return c;
          } else {
            var cx = vcmp(h1.value, h2.value);
            if (cx) {
              return cx;
            } else {
              _e2 = Bs_internalAVLtree.stackAllLeft(h2.right, e2[1]);
              _e1 = Bs_internalAVLtree.stackAllLeft(h1.right, e1[1]);
              continue ;
              
            }
          }
        } else {
          return 0;
        }
      } else {
        return 0;
      }
    };
  } else if (len1 < len2) {
    return -1;
  } else {
    return 1;
  }
}

function eq(s1, s2, eq$1) {
  var len1 = Bs_internalAVLtree.length0(s1);
  var len2 = Bs_internalAVLtree.length0(s2);
  if (len1 === len2) {
    var _e1 = Bs_internalAVLtree.stackAllLeft(s1, /* [] */0);
    var _e2 = Bs_internalAVLtree.stackAllLeft(s2, /* [] */0);
    var eq$2 = eq$1;
    while(true) {
      var e2 = _e2;
      var e1 = _e1;
      if (e1) {
        if (e2) {
          var h2 = e2[0];
          var h1 = e1[0];
          if (h1.key === h2.key && eq$2(h1.value, h2.value)) {
            _e2 = Bs_internalAVLtree.stackAllLeft(h2.right, e2[1]);
            _e1 = Bs_internalAVLtree.stackAllLeft(h1.right, e1[1]);
            continue ;
            
          } else {
            return /* false */0;
          }
        } else {
          return /* true */1;
        }
      } else {
        return /* true */1;
      }
    };
  } else {
    return /* false */0;
  }
}

function ofArray(xs) {
  var result = null;
  for(var i = 0 ,i_finish = xs.length - 1 | 0; i <= i_finish; ++i){
    var match = xs[i];
    result = add(result, match[0], match[1]);
  }
  return result;
}

var empty = Bs_internalAVLtree.empty0;

var isEmpty = Bs_internalAVLtree.isEmpty0;

var singleton = Bs_internalAVLtree.singleton0;

var iter = Bs_internalAVLtree.iter0;

var fold = Bs_internalAVLtree.fold0;

var forAll = Bs_internalAVLtree.forAll0;

var exists = Bs_internalAVLtree.exists0;

var filter = Bs_internalAVLtree.filter0;

var partition = Bs_internalAVLtree.partition0;

var length = Bs_internalAVLtree.length0;

var toList = Bs_internalAVLtree.toList0;

var minBinding = Bs_internalAVLtree.minKVOpt0;

var maxBinding = Bs_internalAVLtree.maxKVOpt0;

var map = Bs_internalAVLtree.map0;

var mapi = Bs_internalAVLtree.mapi0;

var checkInvariant = Bs_internalAVLtree.checkInvariant;

exports.empty = empty;
exports.ofArray = ofArray;
exports.isEmpty = isEmpty;
exports.mem = mem;
exports.add = add;
exports.singleton = singleton;
exports.remove = remove;
exports.merge = merge;
exports.cmp = cmp;
exports.eq = eq;
exports.iter = iter;
exports.fold = fold;
exports.forAll = forAll;
exports.exists = exists;
exports.filter = filter;
exports.partition = partition;
exports.length = length;
exports.toList = toList;
exports.minBinding = minBinding;
exports.maxBinding = maxBinding;
exports.split = split;
exports.findOpt = findOpt;
exports.findAssert = findAssert;
exports.findWithDefault = findWithDefault;
exports.map = map;
exports.mapi = mapi;
exports.checkInvariant = checkInvariant;
/* No side effect */

File: jscomp/others/bs_MapString.ml
# 2 "map.cppo.ml"
type key = string

# 9
module N = Bs_internalAVLtree

type ('key, 'a, 'id) t0 = ('key,'a) N.t0 

type  'a t = (key,'a) N.t0 





let empty = N.empty0      
let isEmpty = N.isEmpty0
let singleton = N.singleton0
let minBinding = N.minKVOpt0
let maxBinding = N.maxKVOpt0
let iter = N.iter0      
let map  = N.map0
let mapi = N.mapi0
let fold = N.fold0
let forAll = N.forAll0
let exists = N.exists0    
let filter = N.filter0
let partition = N.partition0
let length = N.length0
let toList = N.toList0
let checkInvariant = N.checkInvariant

let rec add  t (x : key) (data : _)  = 
  match N.toOpt t with
  | None -> 
    N.(return @@ node ~left:empty ~key:x ~value:data ~right:empty ~h:1)
  | Some n (* Node(l, v, d, r, h) *) ->
    let l,k,v,r = N.(left n, key n, value n, right n) in 
    if x = k then
      N.(return @@ node ~left:l ~key:x ~value:data ~right:r ~h:(h n))
    else if x < k then
      N.(bal (add l x data ) k v r)
    else
      N.(bal l k v (add r x data ))

let rec findOpt (x : key) n =
  match N.toOpt n with 
    None -> None
  | Some n  ->
    let v = N.key n in 
    if x = v then Some (N.value n)
    else findOpt x (if x < v then N.left n else N.right n)

let rec findAssert (x : key) n = 
  match N.toOpt n with 
  | None ->
    [%assert "Not_found"]
  | Some n  ->
    let v = N.key n in 
    if x = v then (N.value n)
    else findAssert x (if x < v then (N.left n) else (N.right n))

let rec findWithDefault n (x : key) def =
  match N.toOpt n with 
  | None -> def    
  | Some n -> 
    let v = N.key n in 
    if x = v then (N.value n)
    else findWithDefault (if x < v then (N.left n) else (N.right n)) x def

let rec mem n (x : key)= 
  match N.toOpt n with 
    None -> false
  | Some n (* Node(l, v, d, r, _) *) ->
    let v = N.key n in 
    x = v || mem (if x < v then N.left n else N.right n) x 

let rec remove n (x : key) = 
  match N.toOpt n with 
  |  None -> n    
  |  Some n ->
    let l,v,r = N.(left n, key n, right n) in 
    if x = v then
      match N.toOpt l, N.toOpt r with
      | None, _ -> r 
      | _, None -> l 
      | _, Some rn -> 
        let kr, vr = ref (N.key rn), ref (N.value rn) in 
        let r = N.removeMinAuxWithRef rn kr vr in 
        N.bal l !kr !vr r 
    else if x < v then
      N.(bal (remove l x ) v (value n) r)
    else
      N.(bal l v (value n) (remove r x ))

let rec splitAux (x : key) (n : _ N.node) : _ t0 * _ option  * _ t0 =  
  let l,v,d,r = N.(left n , key n, value n, right n) in  
  if x = v then (l, Some d, r)
  else     
    if x < v then
      match N.toOpt l with 
      | None -> 
        N.(empty , None, return n)
      | Some l -> 
        let (ll, pres, rl) = splitAux x l in (ll, pres, N.join rl v d r)
    else
      match N.toOpt r with 
      | None ->
        N.(return n, None, empty)
      | Some r -> 
        let (lr, pres, rr) = splitAux x r in (N.join l v d lr, pres, rr)
      

let rec split (x : key) n =
  match N.toOpt n with 
    None ->
    N.(empty, None, empty)
  | Some n -> 
    splitAux x n 

let rec merge s1 s2 f =
  match N.(toOpt s1, toOpt s2) with
    (None, None) -> N.empty
  | Some n (* (Node (l1, v1, d1, r1, h1), _)*), _ 
    when N.(h n >= (match N.toOpt s2 with None -> 0 | Some n -> N.h n)) ->
    let (l1,v1,d1,r1) = N.(left n, key n, value n, right n ) in 
    let (l2, d2, r2) = split v1 s2 in
    N.concatOrJoin (merge l1 l2 f) v1 (f v1 (Some d1) d2 [@bs]) (merge r1 r2 f)
  | (_, Some n) (* Node (l2, v2, d2, r2, h2) *)  ->
    let (l2, v2, d2, r2) = N.(left n, key n, value n, right n) in 
    let (l1, d1, r1) = split v2 s1 in
    N.concatOrJoin (merge l1 l2 f) v2 (f v2 d1 (Some d2) [@bs]) (merge r1 r2 f)
  | _ ->
    assert false

let rec compareAux e1 e2 vcmp =
   match e1,e2 with 
   | h1::t1, h2::t2 ->
    let c = Pervasives.compare (N.key h1 : key) (N.key h2)  in 
    if c = 0 then 
      let cx = vcmp (N.value h1) (N.value h2) [@bs] in 
      if cx = 0 then
          compareAux 
          (N.stackAllLeft  (N.right h1) t1 ) 
          (N.stackAllLeft (N.right h2) t2)
          vcmp 
      else  cx
    else c 
  | _, _ -> 0    

let cmp s1 s2 cmp = 
  let len1, len2 = N.length0 s1, N.length0 s2 in 
  if len1 = len2 then 
    compareAux 
      (N.stackAllLeft s1 []) 
      (N.stackAllLeft s2 []) 
      cmp 
  else if len1 < len2 then -1 
  else 1 


let rec eqAux e1 e2  eq =
    match e1,e2 with 
    | h1::t1, h2::t2 ->
     if (N.key h1 : key) =  (N.key h2)  && 
        eq (N.value h1) (N.value h2) [@bs] then
          eqAux (
            N.stackAllLeft  (N.right h1) t1 ) 
            (N.stackAllLeft (N.right h2) t2)
            eq
      else  false    
    | _, _ -> true (*end *)  

let eq s1 s2 eq =      
  let len1,len2 = N.length0 s1, N.length0 s2 in 
  if len1 = len2 then 
    eqAux 
    (N.stackAllLeft s1 [])
    (N.stackAllLeft s2 []) eq 
  else false  

let ofArray  (xs : _ array) : _ t0 =     
  let result = ref N.empty in 
  for i = 0 to Array.length xs - 1 do  
    let k, v = (Bs_Array.unsafe_get xs i) in 
    result := add !result k v 
  done ;
  !result 




